DT: check vendor prefixes

mbolivar-nordic opened the pull request:
This is based on https://github.com/zephyrproject-rtos/zephyr/pull/35313.

As suggested by @trond-snekvik, extends the checks in the PR it's based on to also emit warnings when a vendor prefix not in vendor-prefixes.txt or a special list of additional exemptions is present in the DTS.

mbolivar-nordic committed:
edtlib: only do compatible checks once

Remember the compatibles we've validated, and don't check them again.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: fix vendor-prefixes.txt whitespace

This line is here to satisfy the commit message linter.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: bindings: update vendor-prefixes.txt

Update our vendor-prefixes.txt file to match the data in upstream
Linux's vendor-prefixes.yaml file as of Linux commit
8ac91e6c6033ebc12c5c1e4aa171b81a662bd70f.

(We can't easily switch to using the Linux YAML file directly because
our documentation tooling depends on this plain text format.)

The changes were done using a script and then cleaned up by hand in
order to not throw away any Zephyr-only prefixes, and to manually
override Linux in some cases.

List of Zephyr-only prefixes, all of which were kept:

    asmedia
    cadence
    gaisler
    lairdconnect
    ovti
    openisa
    particle
    quicklogic
    ruuvi
    segger
    vnd
    weact
    zephyr

Other things to note:

- keep Espressif as 'espressif', overriding Linux's 'esp'
- 'gw' is deprecated in Linux and is unused here; just remove it
- ROCKTECH DISPLAYS LIMITED SHOULD HAVE BEEN ALL CAPS;
  HOW COULD WE POSSIBLY HAVE MISSED THIS BEFORE?
- Did not include '70mai' from upstream, as that violates the legal
  compatible regexp

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: tweak vendor-prefixes.txt format

Use a comment syntax instead of a separator. This is a bit cleaner
and is prep work for moving the parsing code to a more generic place,
namely edtlib.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: move vendor-prefixes.txt parsing in here

Take this helper from the doc tooling and expose it as API.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
edtlib: warn on unrecognized vendors

An unknown vendor prefix is now a warning. We augment the list of
vendor prefixes passed by the user with a grandfathered-in bunch from
Linux.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
cmake: dts: warn on unrecognized vendor prefixes

This uses the list of vendor prefixes provided by the user by default.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: vendor-prefixes.txt: add 'arduino'

This is a common Zephyr addition.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

trond-snekvik commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -475,6 +483,48 @@ def _check(self):
+            _err(f"node '{node.path}' compatible '{compat}' "
+                 'must match this regular expression: '
+                 f"'{compat_re}'")
+
```
Skip the rest if no vendor prefix argument is given?

mbolivar-nordic replied:
I think we do skip the rest if there's no `,`, which means there is no vendor prefix.

We do want to add the compatible to `_checked_compatibles` so that we avoid repeating the check.

trond-snekvik replied:
I think I need to rephrase that a bit😄

I mean, if we don't pass the `--vendor-prefixes` to this script, perhaps this check should be skipped altogether? If I want to run this as part of a downstream CI run now, there's no way to pass this check if I have out-of-tree vendor prefixes, aside from copying the entire Zephyr prefix file

mbolivar-nordic replied:
D'oh, that was my intent but I did it wrong. Should be fixed now; thanks for catching that.

trond-snekvik reviewed (commented):
This looks like it's already closing a lot of gaps, nice!

> should we add a kwarg that makes this warning into an error, for upstream CI?

Aren't warnings treated as errors on this check already? If not, then yes, imo.

Perhaps it should be possible to specify multiple vendor prefix files, so downstream repos that want to run this check can add their local files? If not, there should probably be a way to prevent this check from running.

mbolivar-nordic commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -1,19 +1,22 @@
+#
+# <vendor-prefix><TAB><Full name of vendor>
 
+70mai	70mai Co., Ltd.
```
That's a good point. I can only find one usage in Linux: https://github.com/torvalds/linux/blob/master/arch/arm/boot/dts/mstar-mercury5-ssc8336n-midrived08.dts

I guess that dt-schema would barf on this, which is weird. I'll just drop it from this PR since we don't need it. Thanks.

mbolivar-nordic commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -475,6 +483,48 @@ def _check(self):
+            return
+
+        # The regular expression comes from dt-schema.
+        compat_re = r'^[a-zA-Z][a-zA-Z0-9,+\-._]+$'
```
It's copy/pasted directly from dtschema.

mbolivar-nordic replied:
... well, mostly. I did convert to a raw string.

mbolivar-nordic commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -358,111 +477,155 @@ raspberrypi	Raspberry Pi Foundation
 startek	Startek
 ste	ST-Ericsson
 stericsson	ST-Ericsson
+st-ericsson	ST-Ericsson
```
Jeez, I didn't look at that. I can even find instances of each of them in Linux. What a mess, ha.

mbolivar-nordic commented:
> Aren't warnings treated as errors on this check already? If not, then yes, imo.

DT warnings are kind of a mess. We have a hard twister error on deprecated properties since fbd34dca1e619272a6489de32e7bb4ca9e049fc3, but that is specific to using deprecated properties at all, and doesn't cover this case, which is just making a call to the standard `logging` module.

We could add an edtlib "all_warnings_are_errors" kwarg and set it to True from twister in the right places to cover these things once and for all. Let's see what people think after the merge window opens up again and people have a chance to look at new features like this.

mbolivar-nordic commented:
> Perhaps it should be possible to specify multiple vendor prefix files, so downstream repos that want to run this check can add their local files? If not, there should probably be a way to prevent this check from running.

Yeah, the ergonomics on this for downstream users are a concern for me too. I _think_ that if we change this to automatically skip warns on any compatible we have an explicit binding for in the EDT instance means we don't need multiple vendor prefixes files. Seem reasonable to you?

trond-snekvik reviewed (approved):
This dropped off my radar, but LGTM, apart from the conflicts and the release notes file version. Would these name fixes be considered stable API changes?

mbolivar-nordic commented:
Thanks for review. I've rebased to get fresh CI results, but left the last commit as WIP for now. Will try to return to this for 2.7.

mbolivar-nordic committed:
scripts: rework edtlib warnings-turned-errors

Create a "global" gen_defines.py option and edtlib.EDT constructor
kwarg that turns edtlib-specific warnings into errors. This applies to
edtlib-specific warnings only. Warnings that are just dupes of dtc
warnings are not affected.

Use it from twister to increase DT testing coverage in upstream zephyr.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
> We could add an edtlib "all_warnings_are_errors" kwarg and set it to True from twister in the right places to cover these things once and for all. Let's see what people think after the merge window opens up again and people have a chance to look at new features like this.

I went with this.
